### DISCUSSION TITLE
Recommended list of extras?

### DISCUSSION THREAD

**Discussion by rudenkornk** [LazyVim/LazyVim#3539] [Q&A]
My use case: I would like to delegate as much configuration burden to the maintainers as possible and let them decide which is the best experience as long as they implement it in an efficient lazy-loaded manner, which does not cummulatively slow down the editor. I believe the latter condition is satisfied.

Considering this, I would be happy to get all the latest & greatest:
1. Enable all the functionally, which maintainers think not everybody might like due to ideological reasons, but is considered a UX improvement, even if it potentially breaks backwards compatibility.
2. Do not enable all the legacy/compatibility functionality, which is deperecated and will be removed/unsupported in the future.
3. Do not enable functionality, which is still in its testing phase.

If I get the idea of `extras` correctly, it is an easy way to add some nice but not-universally-accepted-by-everybody functionality to the editor.
Currently there are all three types of extras. For example `lsp.none-ls` as 2-nd type and `trouble-v3` as a 3rd.

Taking into account increasing number of extras, I suppose it would be really nice to see in the documentation (or at least here in discussions) not just a disconnected list, but some categorization or a guide, explaining in short (1) when one want to enable different extras, (2) are there extras confilcts, (3) are there extras functionality duplication (4) simple way to just enable the best UX (even if new version might break compatibility with the old config).

**Comment by dpetka2001** [comment]
> If I get the idea of extras correctly, it is an easy way to add some nice but not-universally-accepted-by-everybody functionality to the editor.
Currently there are all three types of extras. For example lsp.none-ls as 2-nd type and trouble-v3 as a 3rd.

Extras is not some kind of "not-universally-accepted-by-everybody" functionality. The core LazyVim functionality strives to provide a minimal core setup that provides an enjoyable and complete editor usage. Extras just provide additional functionality that not all users might want. For example, the languages Extras provide users with a minimal setup for the specific language, but not all users use the specific language. So, only those who want that will enable the corresponding Extra. Same goes for the plugins. 

`lsp.none-ls` is not deprecated. There has been a forked repo from the original and is maintained by their community. This is included in the Extras, because I have read both here and on Reddit that there are still people who use it and are more comfortable with it. As such it was included in the Extras to give users the choice to use it if they want to.

`trouble-v3` has been merged into main and replaced the old `trouble-v2`, thus it also replaced the Extra as a core plugin and is no longer visible under Extras (except if you're on old LazyVim version).

Extras has different sections of recommended plugins and languages as well as a section with all plugins and languages. `recommended` are shown based on where you open Neovim and taking into consideration mostly similar functionality on how `nvim-lspconfig` does for the LSP servers (`ft` and `root_dir`). 

There might be some conflicts/duplicate functionality between Extras (just yesterday I submitted a PR for disabling Lualine component for symbols when other similar Extras are enabled such as navic or aerial), but usually these get resolved pretty quickly. Only the maintainer has merge rights in this repo and even in his absence LazyVim is still functional and users can incorporate unmerged PRs in their personal configuration pretty easily.

Taking into account everything I said earlier, I really don't see the need for a guide describing what each Extra does in my personal opinion as a simple regular user like you (even I don't use all the Extras and only use those that are useful to me personally). Most Extras are PRs from people that are involved in the specific domain. For example, I couldn't do the Extra for `sql`, because I don't do `sql` myself, since I'm still new to programming myself.

You just enable whatever Extra you want by pressing either `x` on Dashboard or by typing `:LazyExtras` yourself and then just press `x` on the LazyExtras UI to enable the specific Extra you want. Same goes for disabling the Extras. Users should explore the Extras and decide which ones are useful to them. There's no a be-all/end-all recommendation kind of thing that could apply to all users. Everyone's needs and preferences are different.

Everything I said is not meant to discourage you, should you want to create a topic Discussion where you describe what each Extra does.

**Comment by rudenkornk** [comment]
> Extras is not some kind of "not-universally-accepted-by-everybody" functionality. 
> Extras just provide additional functionality that not all users might want.

Well... this is exactly what I said: "not-universally-accepted-by-everybody" functionality. Otherwise it would be just a core part.

> lsp.none-ls is not deprecated. There has been a forked repo from the original and is maintained by their community. This is included in the Extras, because I have read both here and on Reddit that there are still people who use it and are more comfortable with it.

But this is what I am talking about! [Inferring from NEWS.md](https://github.com/LazyVim/LazyVim/blob/51854046dbb0bacf690ddc217cc76bf98c8ec1aa/NEWS.md#L134) I can guess that `conform.nvim` is now considered a primary way of formatting. The reason `none-ls` is still supported is compatibility and opinion of some users. It is OK, I just want to get "the standard" solution, where "the standard" means whatever maintainers think should be a primary way.

> trouble-v3 has been merged

Wow, this is great! This was just an example. I suppose there could be another beta-tested extras.


> There might be some conflicts/duplicate functionality between Extras (just yesterday I submitted a PR for disabling Lualine component for symbols when other similar Extras are enabled such as navic or aerial), but usually these get resolved pretty quickly.

I am afraid this may become messy pretty quickly. Growing number of extras will increase number of potential conflicts quadratically, which will be harder and harder to maintain.

> For example, I couldn't do the Extra for sql, because I don't do sql myself.

All languages are efficiently managed by `nvim-lsp`. I regularly encounter new (to myself) languages and I will be happy if there will be some standard support out-of-the box.

-----

I see your point. Probably my suggestion does not make much sense if we consider extras as blocks, which should be explored by every user.
But this makes extras just another messy plugin ecosystem, which needs to have yet another opinionated wrapper choosing which ones should be enabled or disabled. I do not want to choose between `conform` and `none-ls`. I do not even want to know there is a choice. I want a standard solution providing the best UX without considering any compatibility issues (i.e. I am ok with compatibility breaks).

**Comment by dpetka2001** [comment]
> Well... this is exactly what I said: "not-universally-accepted-by-everybody" functionality. Otherwise it would be just a core part.

Oh I'm sorry about that, seems like I misinterpreted what you meant. Apologies since English is not my native language.

> But this makes extras just another messy plugin ecosystem, which needs to have yet another opinionated wrapper choosing which ones should be enabled or disabled.

I partly agree with this and partly not. It may seem a bit messy, but it's better to have and be able to enable some functionality you want rather than not have at all that option (of course users can just do that in their personal config, I'm just talking about ease of availability and functionality here). 

About the `opinionated` part, in my opinion that is to be expected. Every other Neovim distro that I have checked is in some ways opinionated. And the Extras might also be somewhat opinionated, since they are done by people who have a specific preference about doing things. You can always submit a PR to improve an Extra.

> I do not want to choose between conform and none-ls. I do not even want to know there is a choice.

You might not want that, but other users might. I can understand you might not want to know there is a choice, but I still prefer this way of giving users an option, for those who want that. You are already getting the standard solution from the core LazyVim functionality, which is conform. Extras are just additional choices for users who might want them.

**Comment by rudenkornk** [comment]
> About the opinionated part, in my opinion that is to be expected.

Sure, I am not complaining about this. I understand perfectly that my request (if satisfied) inevitably leads to (1) regular and pretty harsh compatibility breaks and (2) heavily opinionated decisions from maintainers. I am OK with that.

> You are already getting the standard solution from the core LazyVim functionality, which is conform.

For conform -- yes. But there are lots of neat extras like `outline`, `dial`, lots of `dap` extras, `treesitter-context` which need to be enabled explicitly. The problem is not that they are opt-in, but the research one must perform in order to answer several questions before enabling these extras:
1. Is this extra an alternative to something already enabled in the core part?
2. Is this extra an alternative to another extra? If so, which is the better choice?
3. Can this extra conflict with other extras?


I agree that a full guide might be unsuitable solution here. What do you think about marking some extras with smth like `alternative` label?

**Reply by dpetka2001** [reply]
> but the research one must perform in order to answer several questions before enabling these extras

But that is to be expected given that Neovim is a highly configurable editor that the user can tailor depending on his needs, no? 

About the `alternative` label I don't think so. That would mean that someone has to audit every available Extra and know what it does and what alternative functionality another Extra might provide. I'm only familiar with the Extras I use myself and some others maybe. The only ones I could think of `alternative` label are `aerial`, `outline` and `navic` which offer similar functionality with `trouble` (but they could also have differences that some users might prefer, just that I'm not aware of since I don't use them). Of course I could be missing others Extras that I don't use and am not aware of.

In the end, I'm just a regular user like you and just stating my personal opinion and the maintainer has the final say, but I don't believe anything else is actually needed to be honest (but that's just me).

**Reply by rudenkornk** [reply]
> But that is to be expected given that Neovim is a highly configurable editor that the user can tailor depending on his needs, no?

I guess so. But I see here a way to simplify the user's workflow by providing some hints for extras.

> That would mean that someone has to audit every available Extra and know what it does and what alternative functionality another Extra might provide. 

The extras list is formed and moderated by maintainers. I suppose they can easily integrate some "alternatives" hinting into this list just like `recommended` functionality exists now.


> The only ones I could think of alternative label are aerial, outline and navic

There is also [`python` and `python-semshi`], [`Alpha` and `mini-starter`], [`none-ls` and core `conform` & `nvim-lint`].
There are probably others but it really requires research to understand relations between different extras and a core functionality.
